{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "662cf2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(tidyverse))\n",
    "suppressPackageStartupMessages(library(GenomicRanges))\n",
    "suppressPackageStartupMessages(library(Seurat))\n",
    "suppressPackageStartupMessages(library(Signac))\n",
    "suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))\n",
    "suppressPackageStartupMessages(library(EnsDb.Mmusculus.v79))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "59c0a933",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks_path <- \"/fs/nexus-projects/scATAC-seq/piscem/map_output/8k_mouse_cortex_ATACv2_nextgem_Chromium_Controller_fastqs/k_23/m_15/thr=0.7/macs2_peaks.narrowPeak\"\n",
    "mapping_path <- \"/fs/nexus-projects/scATAC-seq/piscem/map_output/8k_mouse_cortex_ATACv2_nextgem_Chromium_Controller_fastqs/k_23/m_15/thr=0.7/map_sorted.bed.gz\"\n",
    "data_name <- \"PBMC\"\n",
    "method_name <- \"piscem\"\n",
    "# genome <- \"mm10\"\n",
    "org <- \"mm10\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e3e342df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# out_path <- \"\"\n",
    "# mapping_cells <- read_tsv(mapping_path, \n",
    "#     col_names=c(\"chr\",\"start\",\"stop\",\"cell\", \"support\"), \n",
    "#     col_types=c(\"-\",\"-\",\"-\",\"-\",\"-\"),\n",
    "#     col_select=\"cell\") %>% \n",
    "#     pull(cell) %>% \n",
    "#     unique()\n",
    "\n",
    "# names(x = mapping_cells) <- paste(data_name, method_name, mapping_cells, sep=\"_\")\n",
    "\n",
    "# mapping_frags <- CreateFragmentObject(path = mapping_path, cells = mapping_cells, max.lines=NULL)\n",
    "\n",
    "# p <- as.data.frame(read.table(peaks_path,header=F,sep=\"\\t\"))\n",
    "# p <- p[,c(1:3)]\n",
    "# colnames(p) <- c(\"chr\",\"start\",\"stop\")\n",
    "# peaks <- suppressPackageStartupMessages(makeGRangesFromDataFrame(p))\n",
    "\n",
    "# mat <- FeatureMatrix(\n",
    "#   fragments = mapping_frags,\n",
    "#   features = peaks,\n",
    "#   process_n = 20000,\n",
    "#   sep = c(\"-\", \"-\"),\n",
    "#   verbose = TRUE\n",
    "# )\n",
    "\n",
    "mapping_assay <- CreateChromatinAssay(mat, fragments = mapping_frags, \n",
    "    genome = org, min.features = 500)\n",
    "seurat_ob <- CreateSeuratObject(mapping_assay, assay = \"peaks\")\n",
    "seurat_ob$Sample <- paste(data_name, method_name, sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8e7de566",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "annotations <- if(org==\"hg38\") GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86) else GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "seqlevelsStyle(annotations) <- 'UCSC'\n",
    "\n",
    "genome(annotations) <- org\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c2ef2523",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating fraction of reads in peaks per cell\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fragmentInfo <- CountFragments(mapping_path)\n",
    "rownames(fragmentInfo) <- paste0(seurat_ob$Sample, \"_\", fragmentInfo$CB)\n",
    "\n",
    "# Attach cell metadata to seurat object\n",
    "seurat_ob$fragments <- fragmentInfo[colnames(seurat_ob), \"frequency_count\"]\n",
    "seurat_ob$mononucleosomal <- fragmentInfo[colnames(seurat_ob), \"mononucleosomal\"]\n",
    "seurat_ob$nucleosome_free <- fragmentInfo[colnames(seurat_ob), \"nucleosome_free\"]\n",
    "seurat_ob$reads_count <- fragmentInfo[colnames(seurat_ob), \"reads_count\"]\n",
    "\n",
    "# Calculate FRiP\n",
    "seurat_ob <- FRiP(\n",
    "  object = seurat_ob,\n",
    "  assay = 'peaks',\n",
    "  total.fragments = \"fragments\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e4a160f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting TSS positions\n",
      "\n",
      "Extracting fragments at TSSs\n",
      "\n",
      "\n",
      "Computing TSS enrichment score\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seurat_ob$blacklist_fraction <- if (org==\"hg38\"){\n",
    "      FractionCountsInRegion(\n",
    "      object = seurat_ob, \n",
    "      assay = 'peaks',\n",
    "      regions = blacklist_hg38\n",
    "        )  \n",
    "    } else {\n",
    "        FractionCountsInRegion(\n",
    "          object = seurat_ob, \n",
    "          assay = 'peaks',\n",
    "          regions = blacklist_mm10\n",
    "        )\n",
    "    }\n",
    "\n",
    "# Compute nucleosome signal score per cell\n",
    "seurat_ob <- NucleosomeSignal(seurat_ob)\n",
    "\n",
    "# Compute TSS enrichment\n",
    "Annotation(seurat_ob) <- annotations\n",
    "seurat_ob <- TSSEnrichment(seurat_ob, fast=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "99f0a89b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seurat_ob_sub <- subset(x = seurat_ob,\n",
    "        subset = nCount_peaks > 1000 &\n",
    "        nCount_peaks < 100000 &\n",
    "        FRiP > 0.15 &\n",
    "        blacklist_fraction < 0.05 &\n",
    "        nucleosome_signal < 4 &\n",
    "        TSS.enrichment > 2\n",
    "        )\n",
    "\n",
    "seurat_ob_sub <- RunTFIDF(seurat_ob)\n",
    "seurat_ob_sub <- FindTopFeatures(seurat_ob_sub, min.cutoff = 'q0')\n",
    "seurat_ob_sub <- RunSVD(seurat_ob_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d13c6923",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\"\n",
      "00:06:53 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "00:06:53 Read 7800 rows and found 29 numeric columns\n",
      "\n",
      "00:06:53 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "00:06:53 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "00:06:53 Writing NN index file to temp file /tmp/RtmpfJDJ2A/file269b9413148653\n",
      "\n",
      "00:06:53 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "00:06:55 Annoy recall = 100%\n",
      "\n",
      "00:06:56 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "00:06:58 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "00:06:59 Commencing optimization for 500 epochs, with 315520 positive edges\n",
      "\n",
      "00:07:07 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seurat_ob_sub <- RunUMAP(object = seurat_ob_sub, reduction = 'lsi', dims = 2:30)\n",
    "seurat_ob_sub <- FindNeighbors(object = seurat_ob_sub, reduction = 'lsi', dims = 2:30)\n",
    "seurat_ob_sub <- FindClusters(object = seurat_ob_sub, verbose = FALSE, algorithm = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ac956d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_ob_sub$nucleosome_group <- ifelse(seurat_ob$nucleosome_signal > 4, 'NS > 4', 'NS < 4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "32ae91a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 1 row containing non-finite outside the scale range (`stat_bin()`).\"\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 2 rows containing missing values or values outside the scale range (`geom_bar()`).\"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC8VBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxsdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgp\nKSkqKiorKyssLCwtLS0uLi4vLy8xMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8\nPDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5P\nT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBh\nYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJz\nc3N0dHR1dXV3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWG\nhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eY\nmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGjo6OkpKSlpaWmpqanp6eoqKipqamqqqqr\nq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9\nvb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7P\nz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh\n4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz\n8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+WEuMiAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAYNElEQVR4nO3de4BcdX338QGCXJRLrfqAFdTWSr304Ym2QnoxxsfH52ET\nIOWSKEGMeCtWrAVRZEVQQAwYBYRISEURpV5AwVIKKCrgDcvlARSByEWgoUITSAhucv7qzF7P\nnCzO7tnP7k7D6/XHOd/ZzO9wspw3zMzO7jYKYMIa030CsDkQEgQICQKEBAFCggAhQYCQIEBI\nEFAzpD32yZ4G/PdWM6Rt9syeBvz3JiQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgRMa0gLBiQOBdNKSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJ\nAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAh\nQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQI\nEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQE\nAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQI6h/TjQ49t7VZ+\ncP7Cs9YPfVBIUNYxpHPf9eFWSOsWnX7fbYefOfRRIUFZx5C+vn5ZK6SrFvYVxY/2Xzv4USFB\n2RieI/WHtKy3uVnTc+vgx4QEZWMN6dQlrXHetc3NibNnz36lkKBkfCHNva65WTJ37tyXTTik\nBW0mejSYXmMN6XPHNTere24b/NjEH9oJic3JWEO6ZsGTRXHt3zwx+DEhQVmnkDasWvXpo1et\n6lu/eMn9t7z1vKEPCwnKOoX0UE+/O4t7PzT/Tcv7hj4sJCibtrcICYnNiZAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBI\nECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJ\nAoQEAUKCACFBgJAgoC2ktU8O7Nf9eEOHZUKCsraQ/vrUgf2qxsoOy4QEZaWQfnb1/3zn1S1X\nndD4UYdlQoKyUkj7N4bt/liHZUKCsvJDu3v+dMH5/b65utMyIUFZ23OkL/5wrMuEBGWVl7/v\n/F7/s6SrH+mwTEhQ1hbSo3829CTp6g7LhARlbSEdu/37zxt4lvRgh2VCgrK2kPb52FiXCQnK\n2kLa79yxLhMSlLWFtOyAsS4TEpS1hdT31kXf+eXdLes6LBMSlLWF9NqGV+2gjraQzvvoCYO8\naRXGw/cjQUBbSHfdPOjGTm+2ExKUtX8/kudIUEtbSBed3vKRfXZZ/nCHZUKCslGfI52/T6dl\nQoKy0V9seO6vOiwTEpSNHtLzr++wTEhQ1hbSmkf63fXRxgMdlgkJykZ/1W5Rp2VCgrK2kD7z\n3pYjP/S1vk7LhARl3tkAAZWQfti7aMF7vvREx2VCgrK2kDa+beAp0kvu6rRMSFDW/s6GxhE/\nW/3Efefv8hedlgkJytpCOuSNA/srGvd2WCYkKGsL6fW9A/sntvhBh2VCgrK2kPY/dGB/Z+Pm\nDsuEBGVtIX1666+3dvfM3tXvR4LxaP9FY3s2dv2r1/7Jllt9tdMyIUFZ+9eRHvvonjtu+4KD\nO/12JCFBO+9sgIBySLdd2r/7/hh+3qqQoKwU0uVb792/f1fjsI7LhARlIyE9uvPuP+4f+t7X\nWNZpmZCgbCSkM0d+A/Nr/3hjh2VCgrKRkA59xfD4xcb9HZYJCcpGQnrDyI8Ouq7R6ZfJCgnK\nRkKa91fD46WN2zssExKUjYR0zHa/GRrfvu3aDsuEBGUjId3QePPgSwyXzzik0zIhQVnp60iL\nG7P/ZV2x8db3z3jOPZ2WCQnKSiGtf3ujseVOz2g09rih4zIhQVnbe+1u/MAbXv26t1zU+Wef\nCAnaeNMqBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAdMT0oKqCR0Npp2QIEBIECAkCBASBAgJAoQE\nAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQ\nIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQ\nEgQICQLGGtKanpbjh24KCcrGGtIDPXesWrVq9dBNIUHZWEP6Rc8T5ZtCgrKxhvTTecsOf8fZ\njw/dFBKUjTWk6xedf9dP3tHbGk+cPXv2K4UEJeN61e7mnvub2yVz58592eYQkoKJGVdIj/bc\nNDhtFg/thETMWEO69kvNzY09qwZvCgnKxhrSTfMu/o9b333S0E0hQdmYH9pd8575i5etG7ol\nJCh7Or9FSEjECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBAS\nBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFC\nggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQI6JKQRr+mJ/lCL/1DNcXECKk6Qg1Cqo5Qg5CqI9QgpOoINQipOkINQqqO\nUIOQqiPUIKTqCDUIqTpCDUKqjlCDkKoj1CCk6gg1CKk6Qg1Cqo5Qg5CqI9QgpOoINQipOkIN\nQqqOUIOQqiPUIKTqCDUIqTpCDUKqjlCDkKoj1CCk6gg1CKk6Qg1Cqo5Qg5CqI9QgpOoINQip\nOkINQqqOUIOQqiPUIKTqCDUIqTpCDUKqjlCDkKoj1CCk6gg1CKk6Qg1Cqo5Qg5CqI9QgpOoI\nNQipOkINQqqOUIOQqiPUIKTqCDUIqTpCDUKqjlCDkKoj1CCk6gg1CKk6Qg1Cqo5QQxeFNHgx\nt1/epV2tf0yns9h0hBqEVB2hBiFVR6hBSNURahBSdYQahFQdoQYhVUeoQUjVEWoQUnWEGoRU\nHaEGIVVHqEFI1RFqEFJ1hBqEVB2hBiFVR6hBSNURahBSdYQahFQdoQYhVUeoQUjVEWoQUnWE\nGoRUHaEGIVVHqEFI1RFqEFJ1hBqEVB2hBiFVR6hBSNURahBSdYQahFQdoQYhVUeoQUjVEWoQ\nUnWEGoRUHaEGIVVHqEFI1RFqEFJ1hBqEVB2hBiFVR6hBSNURahBSdYQahFQdoQYhVUeoQUjV\nEWoQUnWEGoRUHaEGIVVHqEFI1RFqEFJ1hBqEVB2hhu4Mqb2g9ou8/Zr/nQWMrK800/YPLNpv\njXqm7adRO7vxr9tkxYSSf6q/YvBoT1dCEpKQAoQkJCEFCElIQgoQkpCEFCAkIQkpQEhCElKA\nkIQkpAAhCUlIAUISkpAChCQkIQUISUhCChCSkIQUICQhCSlASEISUoCQhCSkACEJSUgBQhKS\nkAKEJCQhBQhJSEIKEJKQhBQgJCEJKUBIQhJSgJCEJKQAIQlJSAFCEpKQAoQkJCEFCElIQgoQ\nkpCEFCAkIQkpQEhCElKAkIQkpAAhCUlIAUISkpAChCQkIQUISUhCChCSkIQUICQhCSlASEIS\nUoCQhCSkACEJSUgBQhKSkAKEJCQhBQhJSEIKEJKQhBQgJCEJKWDMIa384PyFZ60fuiWkun/v\nia4QUncaa0jrFp1+322Hnzl0U0h1/94TXSGk7jTWkK5a2FcUP9p/7eBNIdX9e090hZC601hD\nWtbb3KzpuXXwppDq/r0nukJI3WmsIZ26pLWdd21zs2Tu3LkvE1LNv/dEVwipO40vpLnXFZGQ\nYDMz1pA+d1xzs7rntsGbE3xoB5uZsYZ0zYIni+Lav3li8KaQoGysIa1fvOT+W9563tBNIUHZ\nmL8ge++H5r9ped/QLSFB2fS8RQg2M0KCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQUDek558M\nT3tnTzSk0xsdPW/3LTvfqYs8c/cdpvsUxme3Xab7DMbnObvPmO5TGJftdt+p431ePNGQnvhK\nRwfN/HznO3WRj8w8erpPYXxe1TPdZzA+i2Yum+5TGJeTZh7Z8T6XTjSkMThy5iOTduzJ8O2Z\nX5ruUxifVx063WcwPsfOvHe6T2Fcvjfzc+O4t5CGCGmSCakeIU0yIU2ubgkJnkaEBAFCgoDJ\nC+nXt6+etGMnPfrzB/v3j9z2UNu+e62/6T+b2w13/WJ9Udp3rdW3P9C/H7oguv3CePjnA+c7\nvgtiskJ6/Jj9Fu/7lUk6eNCGT+/3tvlHNf/NLpu3eN+PPTmy72Ireq4rirsWH3TIgp+O7LvW\nP+33tv161w9fEN1+Yaw7/k1HH/q+h8d9QUxWSJ89YnVx4763TtLRc7618J7i0cOXFd8/4I7i\n4cO+PLzvYncsOqgZ0t8u2VhcuHDt8L5b3bDw7uLR0+4YviC6/cL4yuJ1Rd8/LB33BTFZIS28\nqrnpPWuSjp5z0w+bmzM+XHx8aXP/T+8a3nevvr+74pDripU9q5rjgdcM7af7rJ7Sxz//xM+b\n/30fviC6/cI495jmZukJ474gJimk3/Tc2dyuOHpyjh7Wd8QXi3de3Bx+Mve3Q/vpPqenduGx\nRTOk7x7Umt//haH99J7T73DYmYe/b/6SDUMXRNdfGPe++cJ/u+zNt4z7gpikkO7taT1h+/IR\nk3P0rL7T3ruuOOSK5nRrz+qh/XSf1FO6Z+GDrZAuO6x149izh/bTe1K/w77veby4/4B/Hrog\nuv7C2PCFhUctWvL4uC+ISQrpgZ77mtsLjpyco0et/uAH1zT/w/nt5nhTz9qh/XSf1VPZeNQl\nRSukKxa1bh2zfGg/vWf1O+z71eZmyclDF0TXXxhfOOLxou+EE8Z9QUxSSOvn/Vtzu/Tjk3P0\npFVv/2xfc3fU+c3NlQuG993qW++8e+XKhZc89LN5rReS3nbJ0H66z+spLb6wuTnr+KELousv\njPe2HiVfPX/cF8RkvdjwgWXNmt5y2SQdPWfd2y/o33/hvRubT4xPGd53q2WHN81bdPq6+dcW\nxd099wztp/u8ntLSv29u/m7F8AXR7RfGsZ9qbr562LgviMkK6aZ9l19xzLu7+yuFLcsPvbrp\n+8Wjh5xw5dIDfjW872rNh3bFlw7++qWLPzWy71b3H3Tyd09d8JvhC6LbL4wf7nvRTf+88Gvj\nviAm7Z0Ntyw5bsV/TtbBc84+puWEonjo7A8vXVmM7LvZx/9/88nSlScef0nfyL5r3XtG77J/\nL0YuiG6/MG47o/f064txXxDeawcBQoIAIUGAkCBASBAgJAgQEgQICQKENEV6Gzv9fr+lEzvO\nI71D06yXjum+c3avfPSCra4abe3N231yYmf29CakKdLbuC5ynG8M/xvrHFL/fashrdz+A6Ov\n/cyMbv6O9W4npClSCuncf1x77meKou/bn1ryzf439zy8/KRL+i7/SFEsX7HxWycte7h48JxP\nDnzP6/eXnPyN1vvSlq8ovnvqmbcUxfmzG73nDBxmIIZN7jFyuIH7znlhcf0nB/6g32E7PdJc\nu0fxr59Y+ovmzXPO23DpyZ++ozk9udv/nYpPxGZKSFOkFNJf7HXA7+1TPPiyGa+audXLm1f1\nzc/ZZq8/nHPYtkUx+9WL5hy47e7X/PHBf944pSgef+MWe+693R/+svUH7/pfh758xvnFcS9o\n/L/Bb+ZphTTKPUYON3DfOS89bqeZO7b+oN+vtzqqtfbli3f4021nXFgUe+05b+83v2jri5of\nPK1x85R/WjYbQpoipZBm//6iDUXx7saPiuKyRvNx1pztflpsePd2z2xOW5/Y/B9W47n3FBtf\nvUtRHLnlZUVx13P/svUH7ymK9bu9oigWtz20G+0eI4frv++cZ+2/tlj1vFcMrlrR+E5r7Tbz\n1hX37vbcJ4tZW/xDUax5RXMqbm+cNMWflM2IkKZIKaQ5jda7ie++tjXv/IZizZYHNofHdmiF\ntOVjRXFT4y1Fq7O1fTvs27rLRxp3FXO2aP0g9YOfUQlptHuUDjcQUuPXze0BzxhctWjb1gPB\nWY3WNwac2Phec2r9WL9PNqeieP7rJ/mTsBkT0hTpbcxd3O/2Ys6O/R956MLTTjpp51nFjY3j\nWzfntEJq/m+ouLv/9tGNR25vvObopn0bFxdz/kfrLq0w2kIa7R6lww2ENLy03+te2L/2ea3t\n1xr/WMxq/SOLbzQ+39y+Zo/J/ixsvoQ0RXobe8zs95PB19G+ss2z/8+BB20/q7i2saR1+8Bn\nDr7Cdnf/I6xmSD9uvOQN/a4cXLJJSKPdo3S4xSOv2g0ve+Wr+9f+UWt7eeOzxayXDE1F0fPs\nSf80bLaENEXKD+1al/b6Z81sPowrdp1V3NA4sfXR/71JSL9ovK9tyaYhjXaP0uFGDenP+tf2\nJ/O1xvnFrF1b01cbrRcj9hFSbUKaItWQbmm0fgDInVvMKn7TaP0koMd33CSkDTvt1br7qjVP\nGdJo9ygdbrSQXvei/rWN1u8q+ljj+mLWFq3nSB9ttL4p9DUdvzLFUxHSFKmG9O+Ng5qb1z+/\n+bTkVTvdVmz42502Can4+8bniuKRWS/+bamGIxoPDB6m9ardaPcYOVz/fSshHbpt68cOzZqx\nqK944EV/0FfM2ro53feCP2h9QcuLDfUJaYpUQyre2djzdTtf+I7GX/7yu9tt+5oXHrxw05Ae\nf2PjxXs969nXlGu4sPG8PQcOM/B1pE3vMXK4/vtWQlrRaH2pd6+Zb9l17x23+WbzKH9y6C57\n79Caitsb3ftTsrqekKbI1b3Dv/jx86f37y7/xFkrizVnnLO6+NUZp1y5cf7Og3/ySG/rpegr\netc1tz9YcvKXHxtecnHrvXMXn3LRwGGWf6YY/R7Dh+u/b+kPWn69VeuHW5+zovjXU5a2fnpw\n6z0OA1NxWuPGyfwMbN6ENP0e/llr+8qXT83hDvu9tp/hM/Kuuyd3e2PoDJ6OhDT93rrlD4ri\ngtZ7HKbicCu3P7Z8cySkM2bcEDqDpyMhTb8H9thyz5c29l4zRYe7YMZ3SreGQ7p5e99GMQFC\n6gK/veTUT1yxccoOd175d1cMPNNqumCC3yj1NCckCBASBAgJAoQEAUKCACFBgJAgQEgQ8F83\nQ3x9to+bWgAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "FragmentHistogram(object = seurat_ob, group.by = 'nucleosome_group', region = \"chrY-1-2000000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "68191f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (object, assay = NULL, region = \"chr1-1-2000000\", group.by = NULL, \n",
       "<span style=white-space:pre-wrap>    cells = NULL, log.scale = FALSE, ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    cells &lt;- SetIfNull(x = cells, y = colnames(x = object))</span>\n",
       "<span style=white-space:pre-wrap>    assay &lt;- SetIfNull(x = assay, y = DefaultAssay(object = object))</span>\n",
       "<span style=white-space:pre-wrap>    if (!inherits(x = object[[assay]], what = \"ChromatinAssay\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"The requested assay is not a ChromatinAssay.\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    reads &lt;- MultiGetReadsInRegion(object = object, assay = assay, </span>\n",
       "<span style=white-space:pre-wrap>        region = region, cells = cells, verbose = FALSE, ...)</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(x = group.by)) {</span>\n",
       "<span style=white-space:pre-wrap>        groups &lt;- Idents(object = object)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        md &lt;- object[[]]</span>\n",
       "<span style=white-space:pre-wrap>        groups &lt;- object[[group.by]]</span>\n",
       "<span style=white-space:pre-wrap>        groups &lt;- groups[, 1]</span>\n",
       "<span style=white-space:pre-wrap>        names(x = groups) &lt;- rownames(x = md)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    reads$group &lt;- groups[reads$cell]</span>\n",
       "<span style=white-space:pre-wrap>    if (length(x = unique(x = reads$group)) == 1) {</span>\n",
       "<span style=white-space:pre-wrap>        p &lt;- ggplot(data = reads, aes(length)) + geom_histogram(bins = 200)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        p &lt;- ggplot(data = reads, mapping = aes(x = length, fill = group)) + </span>\n",
       "<span style=white-space:pre-wrap>            geom_histogram(bins = 200) + facet_wrap(~group, scales = \"free_y\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    p &lt;- p + xlim(c(0, 800)) + theme_classic() + theme(legend.position = \"none\", </span>\n",
       "<span style=white-space:pre-wrap>        strip.background = element_blank()) + xlab(\"Fragment length (bp)\") + </span>\n",
       "<span style=white-space:pre-wrap>        ylab(\"Count\")</span>\n",
       "<span style=white-space:pre-wrap>    if (log.scale) {</span>\n",
       "<span style=white-space:pre-wrap>        p &lt;- p + scale_y_log10()</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    return(p)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (object, assay = NULL, region = \"chr1-1-2000000\", group.by = NULL, \n",
       "    cells = NULL, log.scale = FALSE, ...) \n",
       "\\{\n",
       "    cells <- SetIfNull(x = cells, y = colnames(x = object))\n",
       "    assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "    if (!inherits(x = object{[}{[}assay{]}{]}, what = \"ChromatinAssay\")) \\{\n",
       "        stop(\"The requested assay is not a ChromatinAssay.\")\n",
       "    \\}\n",
       "    reads <- MultiGetReadsInRegion(object = object, assay = assay, \n",
       "        region = region, cells = cells, verbose = FALSE, ...)\n",
       "    if (is.null(x = group.by)) \\{\n",
       "        groups <- Idents(object = object)\n",
       "    \\}\n",
       "    else \\{\n",
       "        md <- object{[}{[}{]}{]}\n",
       "        groups <- object{[}{[}group.by{]}{]}\n",
       "        groups <- groups{[}, 1{]}\n",
       "        names(x = groups) <- rownames(x = md)\n",
       "    \\}\n",
       "    reads\\$group <- groups{[}reads\\$cell{]}\n",
       "    if (length(x = unique(x = reads\\$group)) == 1) \\{\n",
       "        p <- ggplot(data = reads, aes(length)) + geom\\_histogram(bins = 200)\n",
       "    \\}\n",
       "    else \\{\n",
       "        p <- ggplot(data = reads, mapping = aes(x = length, fill = group)) + \n",
       "            geom\\_histogram(bins = 200) + facet\\_wrap(\\textasciitilde{}group, scales = \"free\\_y\")\n",
       "    \\}\n",
       "    p <- p + xlim(c(0, 800)) + theme\\_classic() + theme(legend.position = \"none\", \n",
       "        strip.background = element\\_blank()) + xlab(\"Fragment length (bp)\") + \n",
       "        ylab(\"Count\")\n",
       "    if (log.scale) \\{\n",
       "        p <- p + scale\\_y\\_log10()\n",
       "    \\}\n",
       "    return(p)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (object, assay = NULL, region = \"chr1-1-2000000\", group.by = NULL, \n",
       "    cells = NULL, log.scale = FALSE, ...) \n",
       "{\n",
       "    cells <- SetIfNull(x = cells, y = colnames(x = object))\n",
       "    assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "    if (!inherits(x = object[[assay]], what = \"ChromatinAssay\")) {\n",
       "        stop(\"The requested assay is not a ChromatinAssay.\")\n",
       "    }\n",
       "    reads <- MultiGetReadsInRegion(object = object, assay = assay, \n",
       "        region = region, cells = cells, verbose = FALSE, ...)\n",
       "    if (is.null(x = group.by)) {\n",
       "        groups <- Idents(object = object)\n",
       "    }\n",
       "    else {\n",
       "        md <- object[[]]\n",
       "        groups <- object[[group.by]]\n",
       "        groups <- groups[, 1]\n",
       "        names(x = groups) <- rownames(x = md)\n",
       "    }\n",
       "    reads$group <- groups[reads$cell]\n",
       "    if (length(x = unique(x = reads$group)) == 1) {\n",
       "        p <- ggplot(data = reads, aes(length)) + geom_histogram(bins = 200)\n",
       "    }\n",
       "    else {\n",
       "        p <- ggplot(data = reads, mapping = aes(x = length, fill = group)) + \n",
       "            geom_histogram(bins = 200) + facet_wrap(~group, scales = \"free_y\")\n",
       "    }\n",
       "    p <- p + xlim(c(0, 800)) + theme_classic() + theme(legend.position = \"none\", \n",
       "        strip.background = element_blank()) + xlab(\"Fragment length (bp)\") + \n",
       "        ylab(\"Count\")\n",
       "    if (log.scale) {\n",
       "        p <- p + scale_y_log10()\n",
       "    }\n",
       "    return(p)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (object, assay = NULL, region = \"chr1-1-2000000\", group.by = NULL, \n",
       "    cells = NULL, log.scale = FALSE, ...) \n",
       "{\n",
       "    cells <- SetIfNull(x = cells, y = colnames(x = object))\n",
       "    assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "    if (!inherits(x = object[[assay]], what = \"ChromatinAssay\")) {\n",
       "        stop(\"The requested assay is not a ChromatinAssay.\")\n",
       "    }\n",
       "    reads <- MultiGetReadsInRegion(object = object, assay = assay, \n",
       "        region = region, cells = cells, verbose = FALSE, ...)\n",
       "    if (is.null(x = group.by)) {\n",
       "        groups <- Idents(object = object)\n",
       "    }\n",
       "    else {\n",
       "        md <- object[[]]\n",
       "        groups <- object[[group.by]]\n",
       "        groups <- groups[, 1]\n",
       "        names(x = groups) <- rownames(x = md)\n",
       "    }\n",
       "    reads$group <- groups[reads$cell]\n",
       "    if (length(x = unique(x = reads$group)) == 1) {\n",
       "        p <- ggplot(data = reads, aes(length)) + geom_histogram(bins = 200)\n",
       "    }\n",
       "    else {\n",
       "        p <- ggplot(data = reads, mapping = aes(x = length, fill = group)) + \n",
       "            geom_histogram(bins = 200) + facet_wrap(~group, scales = \"free_y\")\n",
       "    }\n",
       "    p <- p + xlim(c(0, 800)) + theme_classic() + theme(legend.position = \"none\", \n",
       "        strip.background = element_blank()) + xlab(\"Fragment length (bp)\") + \n",
       "        ylab(\"Count\")\n",
       "    if (log.scale) {\n",
       "        p <- p + scale_y_log10()\n",
       "    }\n",
       "    return(p)\n",
       "}\n",
       "<bytecode: 0x55c2d321d9f8>\n",
       "<environment: namespace:Signac>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "FragmentHistogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0084d838",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "\u001b[1m\u001b[33mError\u001b[39m in `separate()`:\u001b[22m\n\u001b[33m!\u001b[39m Can't extract columns that don't exist.\n\u001b[31m✖\u001b[39m Column `ranges` doesn't exist.\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1m\u001b[33mError\u001b[39m in `separate()`:\u001b[22m\n\u001b[33m!\u001b[39m Can't extract columns that don't exist.\n\u001b[31m✖\u001b[39m Column `ranges` doesn't exist.\nTraceback:\n",
      "1. Signac:::MultiGetReadsInRegion(object = seurat_ob, assay = assay, \n .     region = NULL, cells = cells, verbose = FALSE)",
      "2. GetReadsInRegion(cellmap = cellmap, region = region, tabix.file = tabix.file, \n .     ...)",
      "3. StringToGRanges(regions = region, ...)",
      "4. separate(data = ranges.df, col = \"ranges\", sep = paste0(sep[[1]], \n .     \"|\", sep[[2]]), into = c(\"chr\", \"start\", \"end\"))",
      "5. separate.data.frame(data = ranges.df, col = \"ranges\", sep = paste0(sep[[1]], \n .     \"|\", sep[[2]]), into = c(\"chr\", \"start\", \"end\"))",
      "6. tidyselect::vars_pull(names(data), !!enquo(col))",
      "7. pull_as_location2(loc, n, vars, error_arg = error_arg, error_call = error_call)",
      "8. with_subscript_errors(type = \"pull\", {\n .     i <- vctrs::vec_as_subscript2(i, logical = \"error\", arg = error_arg, \n .         call = error_call)\n .     if (length(i) != 1) {\n .         cli::cli_abort(\"{.arg {error_arg}} must select exactly one column.\", \n .             call = error_call)\n .     }\n .     if (is.numeric(i)) {\n .         vctrs::num_as_location2(i, n = n, negative = \"ignore\", \n .             arg = error_arg, call = error_call)\n .     }\n .     else {\n .         vctrs::vec_as_location2(i, n = n, names = names, arg = error_arg, \n .             call = error_call, )\n .     }\n . })",
      "9. withCallingHandlers(expr, vctrs_error_subscript = function(cnd) {\n .     cnd$subscript_action <- subscript_action(type)\n .     cnd$subscript_elt <- \"column\"\n .     cnd_signal(cnd)\n . })",
      "10. vctrs::vec_as_location2(i, n = n, names = names, arg = error_arg, \n  .     call = error_call, )",
      "11. result_get(vec_as_location2_result(i, n = n, names = names, negative = \"error\", \n  .     missing = missing, arg = arg, call = call))",
      "12. cnd_signal(x$err)",
      "13. signal_abort(cnd)",
      "14. signalCondition(cnd)",
      "15. (function (cnd) \n  . {\n  .     cnd$subscript_action <- subscript_action(type)\n  .     cnd$subscript_elt <- \"column\"\n  .     cnd_signal(cnd)\n  . })(structure(list(message = \"\", trace = structure(list(call = list(\n  .     IRkernel::main(), kernel$run(), handle_shell(), executor$execute(msg), \n  .     tryCatch(evaluate(request$content$code, envir = .GlobalEnv, \n  .         output_handler = oh, stop_on_error = 1L), interrupt = function(cond) {\n  .         log_debug(\"Interrupt during execution\")\n  .         interrupted <<- TRUE\n  .     }, error = .self$handle_error), tryCatchList(expr, classes, \n  .         parentenv, handlers), tryCatchOne(tryCatchList(expr, \n  .         names[-nh], parentenv, handlers[-nh]), names[nh], parentenv, \n  .         handlers[[nh]]), doTryCatch(return(expr), name, parentenv, \n  .         handler), tryCatchList(expr, names[-nh], parentenv, handlers[-nh]), \n  .     tryCatchOne(expr, names, parentenv, handlers[[1L]]), doTryCatch(return(expr), \n  .         name, parentenv, handler), evaluate(request$content$code, \n  .         envir = .GlobalEnv, output_handler = oh, stop_on_error = 1L), \n  .     evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos, \n  .         debug = debug, last = i == length(out), use_try = stop_on_error != \n  .             2L, keep_warning = keep_warning, keep_message = keep_message, \n  .         log_echo = log_echo, log_warning = log_warning, output_handler = output_handler, \n  .         include_timing = include_timing), timing_fn(handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .         envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n  .         message = mHandler))), handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .         envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n  .         message = mHandler)), try(f, silent = TRUE), tryCatch(expr, \n  .         error = function(e) {\n  .             call <- conditionCall(e)\n  .             if (!is.null(call)) {\n  .                 if (identical(call[[1L]], quote(doTryCatch))) \n  .                   call <- sys.call(-4L)\n  .                 dcall <- deparse(call, nlines = 1L)\n  .                 prefix <- paste(\"Error in\", dcall, \": \")\n  .                 LONG <- 75L\n  .                 sm <- strsplit(conditionMessage(e), \"\\n\")[[1L]]\n  .                 w <- 14L + nchar(dcall, type = \"w\") + nchar(sm[1L], \n  .                   type = \"w\")\n  .                 if (is.na(w)) \n  .                   w <- 14L + nchar(dcall, type = \"b\") + nchar(sm[1L], \n  .                     type = \"b\")\n  .                 if (w > LONG) \n  .                   prefix <- paste0(prefix, \"\\n  \")\n  .             }\n  .             else prefix <- \"Error : \"\n  .             msg <- paste0(prefix, conditionMessage(e), \"\\n\")\n  .             .Internal(seterrmessage(msg[1L]))\n  .             if (!silent && isTRUE(getOption(\"show.error.messages\"))) {\n  .                 cat(msg, file = outFile)\n  .                 .Internal(printDeferredWarnings())\n  .             }\n  .             invisible(structure(msg, class = \"try-error\", condition = e))\n  .         }), tryCatchList(expr, classes, parentenv, handlers), \n  .     tryCatchOne(expr, names, parentenv, handlers[[1L]]), doTryCatch(return(expr), \n  .         name, parentenv, handler), withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .         envir, enclos, user_handlers)), warning = wHandler, error = eHandler, \n  .         message = mHandler), withVisible(eval_with_user_handlers(expr, \n  .         envir, enclos, user_handlers)), eval_with_user_handlers(expr, \n  .         envir, enclos, user_handlers), eval(expr, envir, enclos), \n  .     eval(expr, envir, enclos), Signac:::MultiGetReadsInRegion(object = seurat_ob, \n  .         assay = assay, region = NULL, cells = cells, verbose = FALSE), \n  .     GetReadsInRegion(cellmap = cellmap, region = region, tabix.file = tabix.file, \n  .         ...), StringToGRanges(regions = region, ...), separate(data = ranges.df, \n  .         col = \"ranges\", sep = paste0(sep[[1]], \"|\", sep[[2]]), \n  .         into = c(\"chr\", \"start\", \"end\")), separate.data.frame(data = ranges.df, \n  .         col = \"ranges\", sep = paste0(sep[[1]], \"|\", sep[[2]]), \n  .         into = c(\"chr\", \"start\", \"end\")), tidyselect::vars_pull(names(data), \n  .         !!enquo(col)), pull_as_location2(loc, n, vars, error_arg = error_arg, \n  .         error_call = error_call), with_subscript_errors(type = \"pull\", \n  .         {\n  .             i <- vctrs::vec_as_subscript2(i, logical = \"error\", \n  .                 arg = error_arg, call = error_call)\n  .             if (length(i) != 1) {\n  .                 cli::cli_abort(\"{.arg {error_arg}} must select exactly one column.\", \n  .                   call = error_call)\n  .             }\n  .             if (is.numeric(i)) {\n  .                 vctrs::num_as_location2(i, n = n, negative = \"ignore\", \n  .                   arg = error_arg, call = error_call)\n  .             }\n  .             else {\n  .                 vctrs::vec_as_location2(i, n = n, names = names, \n  .                   arg = error_arg, call = error_call, )\n  .             }\n  .         }), withCallingHandlers(expr, vctrs_error_subscript = function(cnd) {\n  .         cnd$subscript_action <- subscript_action(type)\n  .         cnd$subscript_elt <- \"column\"\n  .         cnd_signal(cnd)\n  .     }), vctrs::vec_as_location2(i, n = n, names = names, arg = error_arg, \n  .         call = error_call, ), result_get(vec_as_location2_result(i, \n  .         n = n, names = names, negative = \"error\", missing = missing, \n  .         arg = arg, call = call)), vec_as_location2_result(i, \n  .         n = n, names = names, negative = \"error\", missing = missing, \n  .         arg = arg, call = call), tryCatch(vec_as_location(i, \n  .         n, names = names, arg = arg, call = call), vctrs_error_subscript = function(err) {\n  .         err[[\"subscript_scalar\"]] <- TRUE\n  .         err <<- err\n  .         i\n  .     }), tryCatchList(expr, classes, parentenv, handlers), tryCatchOne(expr, \n  .         names, parentenv, handlers[[1L]]), doTryCatch(return(expr), \n  .         name, parentenv, handler), vec_as_location(i, n, names = names, \n  .         arg = arg, call = call), `<fn>`(), stop_subscript_oob(i = i, \n  .         subscript_type = subscript_type, names = names, subscript_action = subscript_action, \n  .         subscript_arg = subscript_arg, call = call), stop_subscript(class = \"vctrs_error_subscript_oob\", \n  .         i = i, subscript_type = subscript_type, ..., call = call), \n  .     abort(class = c(class, \"vctrs_error_subscript\"), i = i, ..., \n  .         call = call)), parent = c(0L, 1L, 2L, 3L, 4L, 5L, 6L, \n  . 7L, 6L, 9L, 10L, 4L, 12L, 13L, 13L, 15L, 16L, 17L, 18L, 19L, \n  . 13L, 13L, 13L, 23L, 24L, 0L, 26L, 27L, 28L, 28L, 30L, 31L, 32L, \n  . 33L, 32L, 35L, 35L, 37L, 38L, 39L, 40L, 37L, 0L, 43L, 44L, 45L\n  . ), visible = c(TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  . TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  . TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE), namespace = c(\"IRkernel\", \n  . NA, \"IRkernel\", NA, \"base\", \"base\", \"base\", \"base\", \"base\", \"base\", \n  . \"base\", \"evaluate\", \"evaluate\", \"evaluate\", \"evaluate\", \"base\", \n  . \"base\", \"base\", \"base\", \"base\", \"base\", \"base\", \"evaluate\", \"base\", \n  . \"base\", \"Signac\", \"Signac\", \"Signac\", \"tidyr\", \"tidyr\", \"tidyselect\", \n  . \"tidyselect\", \"tidyselect\", \"base\", \"vctrs\", \"vctrs\", \"vctrs\", \n  . \"base\", \"base\", \"base\", \"base\", \"vctrs\", \"vctrs\", \"vctrs\", \"vctrs\", \n  . \"rlang\"), scope = c(\"::\", NA, \"local\", NA, \"::\", \"local\", \"local\", \n  . \"local\", \"local\", \"local\", \"local\", \"::\", \":::\", \"local\", \"local\", \n  . \"::\", \"::\", \"local\", \"local\", \"local\", \"::\", \"::\", \":::\", \"::\", \n  . \"::\", \":::\", \":::\", \"::\", \"::\", \":::\", \"::\", \":::\", \":::\", \"::\", \n  . \"::\", \":::\", \":::\", \"::\", \"local\", \"local\", \"local\", \"::\", \"local\", \n  . \":::\", \":::\", \"::\"), error_frame = c(FALSE, FALSE, FALSE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, TRUE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  . FALSE, FALSE, FALSE, FALSE, FALSE, FALSE)), row.names = c(NA, \n  . -46L), version = 2L, class = c(\"rlang_trace\", \"rlib_trace\", \"tbl\", \n  . \"data.frame\")), parent = NULL, i = \"ranges\", subscript_type = \"character\", \n  .     names = character(0), subscript_action = NULL, subscript_arg = \"!!enquo(col)\", \n  .     rlang = list(inherit = TRUE), call = separate(data = ranges.df, \n  .         col = \"ranges\", sep = paste0(sep[[1]], \"|\", sep[[2]]), \n  .         into = c(\"chr\", \"start\", \"end\")), subscript_scalar = TRUE), class = c(\"vctrs_error_subscript_oob\", \n  . \"vctrs_error_subscript\", \"rlang_error\", \"error\", \"condition\")))",
      "16. cnd_signal(cnd)",
      "17. signal_abort(cnd)"
     ]
    }
   ],
   "source": [
    "cells <- Signac:::SetIfNull(x = NULL, y = colnames(x = seurat_ob))\n",
    "assay <- Signac:::SetIfNull(x = NULL, y = DefaultAssay(object = seurat_ob))\n",
    "reads <- Signac:::MultiGetReadsInRegion(object = seurat_ob, assay = assay, \n",
    "        region = NULL, cells = cells, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "85105089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (object, region, fragment.list = NULL, assay = NULL, \n",
       "<span style=white-space:pre-wrap>    ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    if (inherits(x = object, what = \"Seurat\")) {</span>\n",
       "<span style=white-space:pre-wrap>        assay &lt;- SetIfNull(x = assay, y = DefaultAssay(object = object))</span>\n",
       "<span style=white-space:pre-wrap>        object &lt;- object[[assay]]</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    fragment.list &lt;- SetIfNull(x = fragment.list, y = Fragments(object = object))</span>\n",
       "<span style=white-space:pre-wrap>    if (length(x = fragment.list) == 0) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"No fragment files found\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    res &lt;- data.frame()</span>\n",
       "<span style=white-space:pre-wrap>    for (i in seq_along(along.with = fragment.list)) {</span>\n",
       "<span style=white-space:pre-wrap>        tbx.path &lt;- GetFragmentData(object = fragment.list[[i]], </span>\n",
       "<span style=white-space:pre-wrap>            slot = \"path\")</span>\n",
       "<span style=white-space:pre-wrap>        cellmap &lt;- GetFragmentData(object = fragment.list[[i]], </span>\n",
       "<span style=white-space:pre-wrap>            slot = \"cells\")</span>\n",
       "<span style=white-space:pre-wrap>        tabix.file &lt;- TabixFile(file = tbx.path, index = GetIndexFile(fragment = tbx.path, </span>\n",
       "<span style=white-space:pre-wrap>            verbose = FALSE))</span>\n",
       "<span style=white-space:pre-wrap>        open(con = tabix.file)</span>\n",
       "<span style=white-space:pre-wrap>        reads &lt;- GetReadsInRegion(cellmap = cellmap, region = region, </span>\n",
       "<span style=white-space:pre-wrap>            tabix.file = tabix.file, ...)</span>\n",
       "<span style=white-space:pre-wrap>        res &lt;- rbind(res, reads)</span>\n",
       "<span style=white-space:pre-wrap>        close(con = tabix.file)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    return(res)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (object, region, fragment.list = NULL, assay = NULL, \n",
       "    ...) \n",
       "\\{\n",
       "    if (inherits(x = object, what = \"Seurat\")) \\{\n",
       "        assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "        object <- object{[}{[}assay{]}{]}\n",
       "    \\}\n",
       "    fragment.list <- SetIfNull(x = fragment.list, y = Fragments(object = object))\n",
       "    if (length(x = fragment.list) == 0) \\{\n",
       "        stop(\"No fragment files found\")\n",
       "    \\}\n",
       "    res <- data.frame()\n",
       "    for (i in seq\\_along(along.with = fragment.list)) \\{\n",
       "        tbx.path <- GetFragmentData(object = fragment.list{[}{[}i{]}{]}, \n",
       "            slot = \"path\")\n",
       "        cellmap <- GetFragmentData(object = fragment.list{[}{[}i{]}{]}, \n",
       "            slot = \"cells\")\n",
       "        tabix.file <- TabixFile(file = tbx.path, index = GetIndexFile(fragment = tbx.path, \n",
       "            verbose = FALSE))\n",
       "        open(con = tabix.file)\n",
       "        reads <- GetReadsInRegion(cellmap = cellmap, region = region, \n",
       "            tabix.file = tabix.file, ...)\n",
       "        res <- rbind(res, reads)\n",
       "        close(con = tabix.file)\n",
       "    \\}\n",
       "    return(res)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (object, region, fragment.list = NULL, assay = NULL, \n",
       "    ...) \n",
       "{\n",
       "    if (inherits(x = object, what = \"Seurat\")) {\n",
       "        assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "        object <- object[[assay]]\n",
       "    }\n",
       "    fragment.list <- SetIfNull(x = fragment.list, y = Fragments(object = object))\n",
       "    if (length(x = fragment.list) == 0) {\n",
       "        stop(\"No fragment files found\")\n",
       "    }\n",
       "    res <- data.frame()\n",
       "    for (i in seq_along(along.with = fragment.list)) {\n",
       "        tbx.path <- GetFragmentData(object = fragment.list[[i]], \n",
       "            slot = \"path\")\n",
       "        cellmap <- GetFragmentData(object = fragment.list[[i]], \n",
       "            slot = \"cells\")\n",
       "        tabix.file <- TabixFile(file = tbx.path, index = GetIndexFile(fragment = tbx.path, \n",
       "            verbose = FALSE))\n",
       "        open(con = tabix.file)\n",
       "        reads <- GetReadsInRegion(cellmap = cellmap, region = region, \n",
       "            tabix.file = tabix.file, ...)\n",
       "        res <- rbind(res, reads)\n",
       "        close(con = tabix.file)\n",
       "    }\n",
       "    return(res)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (object, region, fragment.list = NULL, assay = NULL, \n",
       "    ...) \n",
       "{\n",
       "    if (inherits(x = object, what = \"Seurat\")) {\n",
       "        assay <- SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
       "        object <- object[[assay]]\n",
       "    }\n",
       "    fragment.list <- SetIfNull(x = fragment.list, y = Fragments(object = object))\n",
       "    if (length(x = fragment.list) == 0) {\n",
       "        stop(\"No fragment files found\")\n",
       "    }\n",
       "    res <- data.frame()\n",
       "    for (i in seq_along(along.with = fragment.list)) {\n",
       "        tbx.path <- GetFragmentData(object = fragment.list[[i]], \n",
       "            slot = \"path\")\n",
       "        cellmap <- GetFragmentData(object = fragment.list[[i]], \n",
       "            slot = \"cells\")\n",
       "        tabix.file <- TabixFile(file = tbx.path, index = GetIndexFile(fragment = tbx.path, \n",
       "            verbose = FALSE))\n",
       "        open(con = tabix.file)\n",
       "        reads <- GetReadsInRegion(cellmap = cellmap, region = region, \n",
       "            tabix.file = tabix.file, ...)\n",
       "        res <- rbind(res, reads)\n",
       "        close(con = tabix.file)\n",
       "    }\n",
       "    return(res)\n",
       "}\n",
       "<bytecode: 0x55c2d31c1fb0>\n",
       "<environment: namespace:Signac>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Signac:::MultiGetReadsInRegion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f02924",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
